pkg/jobconfig/files.go

```diff
@@ -610,34 +625,64 @@ func MakeRegexFilenameLabel(possibleRegex string) string {
 	return label
 }
 
-// IsGenerated returns true if the job was generated using prowgen
-func IsGenerated(job prowconfig.JobBase) bool {
-	_, generated := job.Labels[LabelGenerated]
-	return generated
+// IsGenerated returns true if the job was generated by the provided Generator
+func IsGenerated(job prowconfig.JobBase, generator Generator) (bool, error) {
+	generatedSelector, err := generatedSelectorFor(generator)
+	if err != nil {
+		return false, err
+	}
+	return generatedSelector.Matches(labels.Set(job.Labels)), nil
+}
+
+func generatedSelectorFor(generator Generator) (labels.Selector, error) {
+	requirement, err := labels.NewRequirement(LabelGenerator, selection.Equals, []string{string(generator)})
+	if err != nil {
+		return nil, err
+	}
+	return labels.NewSelector().Add(*requirement), nil
 }
 
-func isStale(job prowconfig.JobBase) bool {
-	genLabel, generated := job.Labels[LabelGenerated]
-	return generated && genLabel != string(NewlyGenerated)
+func staleSelectorFor(generator Generator, pruneLabels labels.Set) (labels.Selector, error) {
+	ls, err := generatedSelectorFor(generator)
+	if err != nil {
+		return nil, err
+	}
+	notNewlyGenerated, err := labels.NewRequirement(string(generator), selection.NotEquals, []string{string(newlyGenerated)})
+	if err != nil {
+		return nil, err
+	}
+	ls = ls.Add(*notNewlyGenerated)
+	for label, value := range pruneLabels {
+		req, err := labels.NewRequirement(label, selection.Equals, []string{value})
+		if err != nil {
+			return nil, err
+		}
+		ls = ls.Add(*req)
+	}
+	return ls, nil
 }
 
-// Prune removes all prowgen-generated jobs that were not created by preceding
-// calls to GenerateJobs() (that method labels them as "newly generated"). All
-// remaining prowgen-generated jobs will be labeled as simply "generated" and
+// Prune removes all generated jobs of the supplied Generator with values that are NOT newly-generated.
 // Prune() returns the resulting job config (which may even be completely empty).
-func Prune(jobConfig *prowconfig.JobConfig) *prowconfig.JobConfig {
+func Prune(jobConfig *prowconfig.JobConfig, generator Generator, pruneLabels labels.Set) (*prowconfig.JobConfig, error) {
 	var pruned prowconfig.JobConfig
+	staleSelector, err := staleSelectorFor(generator, pruneLabels)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
nit: for review it may be easier to write a function here so that there's no diff in the following code:

```go
isGenerated := func(job JobBase) bool {
    return generatedSelector.Matches(labels.Set(job.Labels))
}

isStale := func(job JobBase) bool {
    return staleSelector.Matches(labels.Set(job.Labels))
}
```

If it's not a hassle, can we try this and see the diff in the algorithm go to 0?

Reply:
Its not going to be no diff as the `isGenerated` case body needs to delete a temporary label now, rather than change it, but I can do that.